pkg/util/watch/README.md

```diff
@@ -0,0 +1,95 @@
+# `util/watch` explainer
+
+Here's a high-level overview of `.../util/watch` - our abstraction for the Kubernetes Watch API.
+
+Our implementation is similar to the [`client-go` watch implementation], with some notable
+differences:
+
+1. client-go's is type-erased; ours is generic
+2. client-go's periodically resyncs when necessary, but does not provide the external control or
+   ordering guarantees necessary to use it. Ours provides explicit control for relisting, with
+   notifications when that relisting is complete.
+3. Ours has some additional features, like support for custom indexes
+4. Ours has better observability, by exposing metrics that describe both the API calls and their
+   results, and the current state of the watch (e.g. whether it's healthy).
+
+[`client-go` watch implementation]: https://pkg.go.dev/k8s.io/client-go/tools/watch
+
+---
+
+## How watching generally works
+
+At a high level, watching _generally_ works by first calling `List` to fetch the current state of
+the items, and then repeatedly calling `Watch` _with the [resource version] from listing_ to stream
+the changes to the objects since the initial fetch. Sometimes there are errors with watching that
+require calling `List` again (e.g. resource version is too old).
+
+[resource version]: https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions
+
+So, our general flow looks like:
+
+```
+for item in List() {
+    update store with item
+}
+stream := Watch()
+loop {
+    // stream events...
+    for item in stream {
+        if item is error {
+            goto relist
+        }
+
+        update store with item
+    }
+    goto rewatch
+
+  relist:
+    for item in List() {
+        update store with item
+    }
+
+  rewatch:
+    stream = Watch()
+}
+```
+
+Note that because resource versions are opaque strings (even if they _happen_ to look a lot like
+monotonically increasing integers), we must start a new `Watch` every time we call `List`, because
+we have no way to tell the ordering of events except in relation to each other.
+
+## The basic interface
+
+Calls to `watch.Watch` internally produce an event stream. These are exposed with
+`watch.HandlerFuncs`, a set of callbacks used on every event — whether that comes from updating the
+store with `List` or the actual `Watch` events themselves.
+
+## Our changes: Relisting
+
+In the scheduler, we needed the ability to associate an incoming Pod with the VirtualMachine that
+owns it. Fundamentally, this is _kind of_ racy if we're using watch: it's possible to get an event
+about the Pod before we get any events about the VirtualMachine.
+
+In order to provide an explicit ordering here, we have a support for externally triggering a
+"relist" — forcing an ordering on the store's contents by calling `List` again.
```

Comment:
> as I was writing this comment, I realized we do — this is a bug and we shouldn't be

opened #672

Reply:
> If we receive an event x from A, then we request relisting for stream B / store β, once the relisting completes, we know that the contents of β are guaranteed to be more recent than the original event x, right?

That's right. etcd provides linearizable reads.

> (concretely: in the scheduler, we receive a Pod (x), then relist, then try to fetch the associated VM from β)

My point was that you can fetch the associated VM from $\beta$ because NeonVM makes the pod<->vm association with a single operation on etcd (hence it's atomic and we have no partial results). Always when you list, you get the more recent version of data, however, the VM object could have not be there if the mentioned operation wasn't atomic. 

Let's rewrite the sentences in lines 73-74 with a precise language.

Reply:
Rewrote a chunk of it in af4e0f434e11ef06227ec17e788bfe9c0e0cede0 — does that work?